portfolio/src/main/java/com/google/sps/data/LinkedListTimeRangeGroup.java

```diff
@@ -0,0 +1,187 @@
+    // this is the first time we add anything to the list,
+    // so simply add the new time range to the list and return.
+    if (allTimeRanges.isEmpty()) {
+      allTimeRanges.add(timeRange);
+      return;
+    }
+    ListIterator<TimeRange> iterator = allTimeRanges.listIterator();
+
+    // This variable represents the latest time range previously exmamined.
+    // Initially, this variable points to the time range we want to add.
+    // As the for loop iterates through all the current time ranges already existing
+    // in allTimeRanges, this lastExaminedTimeRange variable is the time range with which
+    // any existing time range merges with, if necessary.
+    TimeRange lastExaminedTimeRange = timeRange;
+
+    while (iterator.hasNext()) {
+      TimeRange currentRange = iterator.next();
+
+      // If the current range is completely contained by the lastExaminedTimeRange,
+      // we need to remove the current range from the list as lastExaminedTimeRange
+      // will eventually replace it.
+      if (lastExaminedTimeRange.contains(currentRange)) {
+        iterator.remove();
+        continue;
+      }
+
+      // The case for when two time ranges need to be merged.
+      // Similar to the case above, it will eventually be replaced so we can
+      // remove currentRange
+      if (lastExaminedTimeRange.overlaps(currentRange)) {
+        lastExaminedTimeRange = mergeTwoTimeRanges(currentRange, lastExaminedTimeRange);
+        iterator.remove();
+      } else if (currentRange.end().isAfter(lastExaminedTimeRange.end())) {
+        // This is the case that lastExaminedTimeRange and the current time range do not overlap.
+        // If the current range from the original list ends after the last examined time range,
+        // add the time range pointed to by lastExaminedTimeRange to the new list before the current
+        // range,and change the lastExaminedTimeRange to point to the current range.
+        iterator.previous();
+        iterator.add(lastExaminedTimeRange);
+        iterator.next();
+        lastExaminedTimeRange = currentRange;
+      }
+
+      // If current time range is the last element in the allTimeRanges list,
+      // then there are two cases: if we just updated our pointer and are at
+      // the same element then there is no need to add it. However, if
+      // lastExaminedTimeRange does not match, this means lastExaminedTimeRange
+      // actually belongs at the end of the new list.
+      if (!iterator.hasNext()) {
+        if (!currentRange.equals(lastExaminedTimeRange)) {
+          iterator.add(lastExaminedTimeRange);
+        }
+        // Once we, potentially, add the last element we should end the
+        // loop because it is the final range in the process.
+        break;
+      }
+    }
+  }
+
+  /**
+   * Helper method for merging two time ranges. This method is package-private so that it can be
+   * tested.
+   */
+  static TimeRange mergeTwoTimeRanges(TimeRange a, TimeRange b) {
```

Comment:
This method can be a helper method defined in the abstract class.

Reply:
Done.